### ISSUE TITLE
Let custom storage engines know which keys they should listen for

### ISSUE THREAD

**Issue by MichaelBrunn3r** [nanostores/persistent#9]
Some storage backends like Tampermonkey/Greasemonkey only allow you to listen to value changes if you specifically added a listener for a key ([GM_addValueChangeListener(key, listener)](https://www.tampermonkey.net/documentation.php#GM_addValueChangeListener)). To enable storages to sync between Tabs, a custom storage engine would need to add a listener for every possible key . This is because currently the storages do not tell the engine which key they want to listen to.

Querying [GM_listValues](https://www.tampermonkey.net/documentation.php#GM_listValues) would be a possible solution, but adding listeners for all values if you just need one seems excessive. Also, if the key does not yet exist (i.e. first time the code is run) or new keys are added during runtime, this solution fails.

As I understand `PersistentEvents:addEventListener` cannot be changed, so what would be a good solution here? Creating a custom `createPersistentStore` function?

**Comment by ai** [comment]
We do not need to be insane about the size. We can change engine contract.

But of course, `createPersistentMap` will not be supported in that environment.

Can I ask you new create minimalistic API suggestion?

**Comment by MichaelBrunn3r** [comment]
Well, currently `createPersistentStore` and `createPersistentMap` are designed around having `window` with `window.addEventListener(eventType, listener, ...)`  and `window.removeEventListener(eventType, listener, ...)` as the default event engine.
`addEventListener` calls always include the `"storage"` event type. In my opinion this is too specific to having `window` as the events engine. Other engines will not have the `"storage"` event type.
Also, listeners first have the check wether the event applies to them. But I think this is outside the scope of this discussion, as `PersistentEvent` already includes a `key` property.
```ts
function listener(e: PersistentEvent) {
  // Listener has to check wether the event applies to it
  if (e.key === name) {
    store.set(decode(e.newValue))
  }
}
...
eventsEngine.addEventListener('storage', listener) // first argument is always "storage", as default events engine is window
```

My proposal would be to change `PersistentEvents` to:
```ts
interface PersistentEvents {
  addEventListener(key: string, listener: PersistentListener): void
  removeEventListener(key: string, listener: PersistentListener): void
}
```

This way very little code has to change and the interopability with `window` can be achieved by a simple adapter:
```ts
// Adapts new API to window
eventsEngine = {
  addEventListener(key, listener): window.addEventListener("storage", listener),
  removeEventListener(key, listener): window.removeEventListener("storage", listener)
}
```

**Comment by MichaelBrunn3r** [comment]
I'm currently trying to implement a proposal. It seems that it still has some issues with `createPersistentMap`, as it creates new keys quite differently. I will write back when I found a solution

**Comment by ai** [comment]
Looks beautiful. Please send PR to save your name in the project history.

I will release it in major update, so do now worry about breaking API.

**Comment by ai** [comment]
Released in 0.4.
